pkg/backends/newrelic/newrelic.go

```diff
@@ -0,0 +1,344 @@
+	n := v.Sub(key)
+	if n == nil {
+		n = viper.New()
+	}
+	return n
+}
+
+// NewClientFromViper constructs a newrelic backend
+func NewClientFromViper(v *viper.Viper) (gostatsd.Backend, error) {
+	nr := getSubViper(v, "newrelic")
+	nr.SetDefault("address", "localhost:8001")
+	nr.SetDefault("event_type", "StatsD")
+	nr.SetDefault("tag_prefix", "")
+	nr.SetDefault("metric_name", "metric_name")
+	nr.SetDefault("metric_type", "metric_type")
+	nr.SetDefault("metric_per_second", "metric_per_second")
+	nr.SetDefault("metric_value", "metric_value")
+	nr.SetDefault("timer_value", "samples_count")
+	nr.SetDefault("timer_min", "samples_min")
+	nr.SetDefault("timer_max", "samples_max")
+	nr.SetDefault("timer_mean", "samples_mean")
+	nr.SetDefault("timer_median", "samples_median")
+	nr.SetDefault("timer_std_dev", "samples_std_dev")
+	nr.SetDefault("timer_sum", "samples_sum")
+	nr.SetDefault("timer_sum_square", "samples_sum_squares")
+
+	return NewClient(
+		nr.GetString("address"),
+		nr.GetString("event_type"),
+		nr.GetString("tag_prefix"),
+		nr.GetString("metric_name"),
+		nr.GetString("metric_type"),
+		nr.GetString("metric_per_second"),
+		nr.GetString("metric_value"),
+		nr.GetString("timer_value"),
+		nr.GetString("timer_min"),
+		nr.GetString("timer_max"),
+		nr.GetString("timer_mean"),
+		nr.GetString("timer_median"),
+		nr.GetString("timer_std_dev"),
+		nr.GetString("timer_sum"),
+		nr.GetString("timer_sum_square"),
+		gostatsd.DisabledSubMetrics(v),
+	)
+}
+
+// NewClient constructs a newrelic backend.
+func NewClient(address, eventType, tagPrefix,
+	metricName, metricType, metricPerSecond, metricValue,
+	timerValue, timerMin, timerMax, timerMean, timerMedian, timerStdDev, timerSum, timerSumSquares string,
+	disabled gostatsd.TimerSubtypes) (*Client, error) {
+
+	return &Client{
+		address:          address,
+		eventType:        eventType,
+		tagPrefix:        tagPrefix,
+		metricName:       metricName,
+		metricType:       metricType,
+		metricPerSecond:  metricPerSecond,
+		metricValue:      metricValue,
+		timerValue:       timerValue,
+		timerMin:         timerMin,
+		timerMax:         timerMax,
+		timerMean:        timerMean,
+		timerMedian:      timerMedian,
+		timerStdDev:      timerStdDev,
+		timerSum:         timerSum,
+		timerSumSquares:  timerSumSquares,
+		disabledSubtypes: disabled,
+	}, nil
+}
+
+// SendMetricsAsync prints the metrics in a MetricsMap for newrelic, preparing payload synchronously but doing the send asynchronously.
+func (client Client) SendMetricsAsync(ctx context.Context, metrics *gostatsd.MetricMap, cb gostatsd.SendCallback) {
+	statsdMetrics := client.preparePayload(metrics, &client.disabledSubtypes)
+	go func() {
+		cb([]error{client.sendPayload(statsdMetrics, &client.disabledSubtypes)})
+	}()
+}
+
+func (client Client) sendPayload(metricSets []Metric, disabled *gostatsd.TimerSubtypes) (retErr error) {
+	for ms := range metricSets {
+
+		go func(ms int) {
```

Comment:
That means `cb()` above is invoked almost immediately, but it should be invoked when all the data has been flushed (or errors encountered). And any errors should be passed to it.